File: tyhp-lang/phpstorm/src/unitTest/kotlin/com/tyhp/lang/workspace/TyhpJsonIncludesTest.kt
package com.tyhp.lang.workspace

import kotlin.test.Test
import kotlin.test.assertEquals

class TyhpJsonIncludesTest {
    @Test
    fun `reads include and exclude arrays`() {
        val globs = parseTyhpJsonGlobs(
            """
            {
                "include": ["./tyhp_src/**/*.tyhp", "../../php-extensions/php8.2.9/**/*.tyhpdef"],
                "exclude": ["./skip/**"]
            }
            """.trimIndent(),
        )
        assertEquals(
            listOf("./tyhp_src/**/*.tyhp", "../../php-extensions/php8.2.9/**/*.tyhpdef"),
            globs.include,
        )
        assertEquals(listOf("./skip/**"), globs.exclude)
    }

    @Test
    fun `missing include is empty`() {
        assertEquals(emptyList(), parseTyhpJsonGlobs("""{ "exclude": [] }""").include)
    }

    @Test
    fun `invalid JSON is treated as empty include`() {
        assertEquals(emptyList(), parseTyhpJsonGlobs("{ not json").include)
    }

    @Test
    fun `non-object JSON is treated as empty include`() {
        assertEquals(emptyList(), parseTyhpJsonGlobs("[]").include)
        assertEquals(emptyList(), parseTyhpJsonGlobs("null").include)
        assertEquals(emptyList(), parseTyhpJsonGlobs("\"x\"").include)
    }
}

File: tyhp-lang/vscode/src/workspace/tyhpJsonIncludes.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import { parseTyhpJsonGlobs } from "./tyhpJsonIncludes";

test("reads include and exclude arrays", () => {
    const globs = parseTyhpJsonGlobs(`{
        "include": ["./tyhp_src/**/*.tyhp", "../../php-extensions/php8.2.9/**/*.tyhpdef"],
        "exclude": ["./skip/**"]
    }`);
    assert.deepEqual(globs.include, [
        "./tyhp_src/**/*.tyhp",
        "../../php-extensions/php8.2.9/**/*.tyhpdef",
    ]);
    assert.deepEqual(globs.exclude, ["./skip/**"]);
});

test("missing include is empty (owns nothing)", () => {
    const globs = parseTyhpJsonGlobs(`{ "exclude": [] }`);
    assert.deepEqual(globs.include, []);
    assert.deepEqual(globs.exclude, []);
});

test("invalid JSON is treated as empty include", () => {
    const globs = parseTyhpJsonGlobs("{ not json");
    assert.deepEqual(globs.include, []);
    assert.deepEqual(globs.exclude, []);
});

test("non-object JSON is treated as empty include", () => {
    assert.deepEqual(parseTyhpJsonGlobs("[]").include, []);
    assert.deepEqual(parseTyhpJsonGlobs("null").include, []);
    assert.deepEqual(parseTyhpJsonGlobs('"x"').include, []);
});

test("drops blank include entries", () => {
    const globs = parseTyhpJsonGlobs(`{ "include": ["", "  ", "./src/**/*.tyhp", 1] }`);
    assert.deepEqual(globs.include, ["./src/**/*.tyhp"]);
});
